src/Freya.Integration/Integration.fs

```diff
@@ -0,0 +1,156 @@
+                        // Convert to FreyaState
+                        let s = { Environment = e
+                                  Meta = { Memos = Map.empty } }
+                        // Execute the pipeline
+                        do! pipeline s |> Async.Ignore
+                    else return () }
+                |> Async.StartAsTask :> Task
+            OwinAppFunc app)
+
+    /// Converts a Freya.Pipeline to an OWIN MidFunc.
+    [<CompiledName("FromFreyaWrapped")>]
+    let ofFreyaWrapped (before: Freya<FreyaPipelineChoice>) (after: Freya<FreyaPipelineChoice>) =
+        OwinMidFunc(fun next ->
+            let app e =
+                // Convert to FreyaState
+                let s = { Environment = e
+                          Meta = { Memos = Map.empty } }
+                async {
+                    // Execute the before pipeline
+                    let! choice, s' = before s
+                    match choice with
+                    // Execute the next OwinAppFunc
+                    | Next ->
+                        let task = next.Invoke(s'.Environment)
+                        do! task.ContinueWith<unit>(fun _ -> ()) |> Async.AwaitTask
+                        // Check the result. If the Task is complete run the after pipeline.
+                        if (task.Status = TaskStatus.RanToCompletion) then
+                            // Execute the after pipeline
+                            do! after s' |> Async.Ignore
+                        else return ()
+                    // Complete the Task
+                    | Halt -> return () }
+                |> Async.StartAsTask :> Task
+            OwinAppFunc app)
+
+    /// Splits a MidFunc into a before and after Freya.Pipeline.
+    [<CompiledName("SplitIntoFreya")>]
+    let splitIntoFreya (midFunc: OwinMidFunc) : Freya<FreyaPipelineChoice> * Freya<FreyaPipelineChoice> =
+        // TODO: find a better way than using TaskCompletionSource instances
+        let nextWasRun = ref false
+        let signalCh = TaskCompletionSource<unit>()
+        let continueCh = TaskCompletionSource<unit>()
+        let midFuncCh = ref Unchecked.defaultof<Task<unit>>
+        let nextSignal =
+            OwinAppFunc(fun _ ->
+                nextWasRun := true
+                signalCh.SetResult ()
+                continueCh.Task :> Task)
+        let before s =
+            async {
+                let! token = Async.CancellationToken
+                // Apply and mutate the OwinEnvironment asynchronously.
+                midFuncCh := midFunc.Invoke(nextSignal).Invoke(s.Environment).ContinueWith<unit>((fun _ -> ()), token)
+                // If haltCh completes first, nextWasRun will be false, and nothing further is necessary.
+                // Otherwise, we'll continue on and call back into the midFunc in the after pipe.
+                let! _ = Async.AwaitTask <| Task.WhenAny([| signalCh.Task; !midFuncCh |])
```

Comment:
Using `Task.WhenAny` to await either a called `next` or immediate handling. This requires .NET 4.5 in this project.